More algorithms to find scrap match

tixomirof opened the pull request:
As I stated in issue #14, I needed more algorithms to speed up scrap combination search in lategame. This pull request provides my own solution to this issue.

With this PR, `sell amount` and `sell quota` will support algorithm flags, such as `sell quota -a:2`. Currently there are only three algorithms:
- DefaultScrapMatchAlgorithm (yours one) - **sell quota -a:1** or **sell quota** with no flags.
- BruteForceMatchAlgorithm (brute forces almost every possible combination until finds one using greedy algorithm) - **sell quota -a:2**
- SuperFastScrapMatchAlgorithm (finds absolutely any combination of scrap that fulfills or exceeds the quota) - **sell quota -a:3**

To implement a new algorithm you will need to create a class (preferably in Helpers.ScrapMatchAlgorithms folder), inherit `BaseScrapMatchAlgorithm` and implement abstract methods (give the algorithm a flag index (don't repeat it) and write your logic in abstract overriden method `RunScrapMatchAlgorithm(int totalScrapValue)`)

Tested this only in singleplayer yet, everything seems to work fine

You can request any changes, I'll do my best to fix everything!

tixomirof committed:
Base scrap match algorithm

Implemented an abstract class for algorithms that find a suitable combination to complete the quota.

tixomirof committed:
feat: Scrap match algorithm implementations

Implemented 3 scrap match algorithms for BaseScrapMatchAlgorithm class, where 1st one is the one from Zehs, 2nd one is brute force and 3rd one is super fast but inaccurate.

tixomirof committed:
feat: Algorithm support for commands

Now all sell commands support algorithms flags (-a:<flagIndex>).
Improved existing code a little: now instead of many arguments to ScrapHelper.GetScrapToSell it only uses a single argument SellCommandRequest, which is a class that contains needed values. Delegated GetScrapToSell logic to this class too.

tixomirof committed:
fix: removed logging

Removed logging I used to quickly debug implemented algorithms

ZehsTeam commented:
Everything looks good so far but I have noticed one issue. The `-o` command flag (with overtime bonus) no longer works when using the sell amount command.

tixomirof committed:
fix: sell with overtime bonus didnt work

flag -o did not do anything and the code kept calculating the exact amount to sell

tixomirof commented:
Thanks for your feedback! The problem was in just a single character :D Now fixed

![26-01-2025_160941597](https://github.com/user-attachments/assets/095687f2-65e4-46dc-a0c4-1c93eef838e0)

ZehsTeam commented:
Awesome! Everything works great now.

ZehsTeam commented:
This has been implemented in the latest Thunderstore release of [SellMyScrap](https://thunderstore.io/c/lethal-company/p/Zehs/SellMyScrap/) v1.11.0
